[DFC] match taxons to dfc product types

rioug opened the pull request:
ℹ️ **This is a funded feature.** Please use the Clockify project with the following name when working on this, including review and test: #9170 OFN DFC Products
#### What? Why?

- Closes [#10809](https://github.com/openfoodfoundation/openfoodnetwork/issues/10809)

Upstream patch has been merged https://github.com/datafoodconsortium/connector-codegen/pull/10, which fix previously blocking issue: https://github.com/datafoodconsortium/connector-ruby/issues/13. We'll have to remove the parser monkey patching once the changes are releases.

To make matching taxon to DFC product type, we added a "dfc id" field so the user can specify which DFC product type a given taxon refers to. "dfc id" are meant to refers to a DFC product type URI defined in https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf, you can browse the file [here](https://showvoc.datafoodconsortium.org/showvoc/#/datasets/ProductTypes/data) . For instance, if you wanted to link the "Bread" category to the "DFC Bread", you would use the following URI : https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#bread
The "dfc id" is then used to pick the correct DFC product type when the API returns a product.
Similarly, when creating/updating a product, the DFC api will try to find the appropriate taxon using the taxon's dfc id.

##### Technical note:
Product types (and other SKOSConcepts) are modelled as Class method on the connector, which return a `SKOSConcept` object. IE: to access the "soft drink" product type you call `connector.PRODUCT_TYPES.DRINK.SOFT_DRINK`. For ease of retrieving the SKOSConcept we want, we first parse the Product type tree and store it in  `SuppliedProductBuilder::PRODUCT_TYPES` hash, ~~with the path to access it, for instance :
 `PRODUCT_TYPES.DRINK.SOFT_DRINK.LEMONADE` is stored as : 
  `SuppliedProductBuilder::PRODUCT_TYPES = { lemonade: [:DRINK, :SOFT_DRINK] }`~~


#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

As a super admin 
- visit Configuration -> Taxonomies
- Edit the product Taxon
- Right click on your chosen taxon -> Edit

NOTE "dfc name" has been renamed "dfc id"

![taxon_dfc_name](https://github.com/openfoodfoundation/openfoodnetwork/assets/40413322/a46c4a98-cdb3-424b-9d35-50c4e658c95c)

- Add a dfc id click update
  -> dfc id is saved
- Right click on the same taxon -> Edit
- Update the dfc id 
  -> dfc id is saved
 

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [x] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.

rioug commented:
As explained here https://github.com/datafoodconsortium/connector-ruby/issues/13 , only the narrowest (more specific one) product type is modelled as a `DataFoodConsortium::Connector::SKOSConcept`, other broader product types are modelled as `ataFoodConsortium::Connector::SKOSInstance`. 
The `DfcLoader.connector.export` requires a `DataFoodConsortium::Connector::SKOSConcept` for the property `dfc-b:hasType`, meaning we can only use the most specific produc type: 
* we can't use Drink -> Soft Drink, 
* we can only use the most specific one Drink -> Soft Drink -> Lemonade 

We also need  `broaders`, `narrowers` and `prefLabels` to be populated for `DataFoodConsortium::Connector::SKOSConcept` so we can easily traverse the product type try for matching purposes

As a starting point this is where the `DataFoodConsortium::Connector::SKOSInstance` is added to the parent, so we can call something like : `DfcLoader.connector.PRODUCT_TYPES.DRINK`
https://github.com/datafoodconsortium/connector-ruby/blob/330d97f0a94696ed51919231e8678e05d67f0021/lib/datafoodconsortium/connector/skos_parser.rb#L126C1-L128

mkllnk commented on db/migrate/20231027041224_add_dfc_name_to_spree_taxons.rb:
```diff
@@ -0,0 +1,5 @@
+class AddDfcNameToSpreeTaxons < ActiveRecord::Migration[7.0]
+  def change
+    add_column :spree_taxons, :dfc_name, :string
```
I imagined a new table that can attach a DFC link to anything. But it may be good to start simple like this. We can always generalise later and move the data into a new table.

mkllnk reviewed (changes requested):
This is great!

I'm tempted to approve but I think that we should change to using a URI instead of a name. That will avoid guessing and allows for name changes.

I also have a pull request on the connector which will replace the guessing of setter method names. I wonder if we need to tweak the import feature there as well:

* https://github.com/datafoodconsortium/connector-codegen/pull/12

rioug commented:
> I also have a pull request on the connector which will replace the guessing of setter method names. I wonder if we need to tweak the import feature there as well:

I had a quick look, and I think it should work as it is but I'll double check that.

mkllnk commented on db/schema.rb:
```diff
@@ -924,6 +925,7 @@
     t.boolean "show_api_key_view", default: false, null: false
     t.string "provider"
     t.string "uid"
+    t.datetime "terms_of_service_accepted_at"
```
```suggestion
```

rioug commented on engines/dfc_provider/app/services/supplied_product_builder.rb:
```diff
@@ -51,9 +53,68 @@ def self.apply(supplied_product, variant)
+    Spree::Taxon.find_by(dfc_id: )
+  end
+
+  private_class_method :product_type, :populate_product_types, :record_type, :call_dfc_product_type,
+                       :taxon
```
indeed, I have been thinking about moving all the matching related code to it's own class. I'll look at it if I have a bit of time

mkllnk commented on engines/dfc_provider/app/services/dfc_product_type_factory.rb:
```diff
@@ -0,0 +1,63 @@
+    type = call_dfc_product_type(current_stack)
+
+    id = type.semanticId
+    @product_types[id] = current_stack
```
Why don't you store the type directly? Then you don't need to resolve the stack when looking up types by id.

rioug replied:
I am not sure to be honest, it makes so much sense now that you are pointing it out :smile:

mkllnk reviewed (approved):
Good refactor.

Am I right to assume that this will become easier when the connector has broader and narrower defined?

rioug commented:
> Am I right to assume that this will become easier when the connector has broader and narrower defined?

With the patched skos parser, skos concept now have broader and narrower populated, but they return an array of URI. So it's just easier to do something like this `DfcLoader.connector.PRODUCT_TYPES.DRINK.SOFT_DRINK.method(false)` to get a list of narrower product types, because you can then call the result straight away without having to parse/transform the URI to get the matching method name. 
The way they implemented it looks like what Maxime was talking about here for the type script version : https://github.com/datafoodconsortium/connector-codegen/pull/10#pullrequestreview-1776049941 , but I wonder if it would be easier to have some sort of store like the PHP version, so then we could just query the connector like : `DfcLoader.connector.store.fetch("https://test.com/productTypes.rdf#lemonade")` so we don't have to do the extra matching work.

But that means your suggested improvement about checking broader terms should be easily feasible.

mkllnk commented on engines/dfc_provider/app/services/dfc_product_type_factory.rb:
```diff
@@ -0,0 +1,61 @@
+    end
+  end
+
+  def record_type(stack, product_type)
+    name = product_type.to_s
+    current_stack = stack.dup.push(name)
+
+    type = call_dfc_product_type(current_stack)
+
+    id = type.semanticId
+    @product_types[id] = type
```
What's the difference between `type` and `product_type`? I was just wondering if we can get rid of `call_dfc_product_type` altogether.

rioug replied:
`type` is the actual object representing a product type, and `product_type` is the name of the method to call to get the actual object. 
I had another look and you can indeed get rid of `call_dfc_product_type` see : [de95b0c](https://github.com/openfoodfoundation/openfoodnetwork/pull/11817/commits/de95b0c7baf636da7dceaa380d671fbe5e490d06)

I am not sure why I though we needed to record the path, I think I just spent to long working on it. It's nice to get an external perspective :pray:

mkllnk reviewed (approved):
Nice improvement. :+1:

rioug commented on engines/dfc_provider/lib/data_food_consortium/connector/importer.rb:
```diff
@@ -120,18 +120,30 @@ def resolve_object(object)
         ).sub(
           "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/measures.rdf#",
           "dfc-m:"
+        ).sub(
+          "dfc-pt:",
+          "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#"
```
@mkllnk I fixed the importer based on your comment here : https://github.com/datafoodconsortium/connector-codegen/pull/12#discussion_r1440030320

mkllnk reviewed (approved):
Great!

Now we just have to reconcile this with the code generator... :upside_down_face:

drummer83 commented:
Hi @rioug,
I've tried testing this PR but it's failing in the early stages already.

## Staging FR
After staging I can't even access the edit page of the taxon anymore (error 500), e.g. `https://staging.coopcircuits.fr/admin/taxonomies/6/taxons/66/edit`. Staging `master` again, this works fine.
Here is the output of the browser console:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/6fe7897e-91e4-4d22-8e8e-e437ffa7b274)
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/715a8f0e-c283-4182-8fc9-cf5e217233b3)

## Staging UK
Staging is failing. Maybe because this server is not configured to work with DFC? Or is it something else?
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/5a670d3f-c7d1-4780-a5de-d9e68ccf09bb)

## Question
Looking at the issue #10809 I wonder if this PR (#11817) is really going to fully close it. The issue is about exporting products (incl. their taxon) to DFC, while testing this PR is asking only to add/edit the new DFC_name field in OFN configuration. Is this really all to fully close #10809?

## Conclusion
I'll move this back to In Dev so we can find the problem.

Thanks for looking into this! :pray:

rioug commented:
Thanks @drummer83 , I fixed the issue, I forgot to update the admin page when I renamed "dfc name" to "dfc id" :facepalm:. I also updated the PR description, it was missing some context.

>  Looking at the issue https://github.com/openfoodfoundation/openfoodnetwork/issues/10809 I wonder if this PR (https://github.com/openfoodfoundation/openfoodnetwork/pull/11817) is really going to fully close it. The issue is about exporting products (incl. their taxon) to DFC, while testing this PR is asking only to add/edit the new DFC_name field in OFN configuration. Is this really all to fully close https://github.com/openfoodfoundation/openfoodnetwork/issues/10809?

The exporting product part is all covered by the automatic testing. If you really want to do some manual testing you can have a look at the api docs for DFC : "<staging_server_url>/api-docs/index.htm".
* set up a product belonging to category that has "dfc_id" set to an existing dfc product type URI 
From the PR description : 
> To make matching taxon to DFC product type, we added a "dfc id" field so the user can specify which DFC product type a given taxon refers to. "dfc id" are meant to refers to a DFC product type URI defined in https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf, you can browse the file [here](https://showvoc.datafoodconsortium.org/showvoc/#/datasets/ProductTypes/data) . For instance, if you wanted to link the "Bread" category to the "DFC Bread", you would use the following URI : https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#bread
* Click on the "SuppliedProducts" GET example 
* Click on the "Try it out" button
* Provide the enterprise id and product id, and click on execute
* You should see something like this in the server response:   "dfc-b:hasType": "dfc-pt:bread"

Let me know if anything isn't clear.

drummer83 commented:
Hi @rioug,
There's a conflict here. Could you please have a look?
Thanks! :pray:

drummer83 commented:
Hi @rioug,
Thanks for fixing the conflict and providing such detailed testing information. 

## Before staging
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/e2febc5c-b6bf-45d9-a4bf-759f1b21b33e)

## After staging
We now have the field DFC-URI available - note that it's neither called `DFC ID` nor `DFC Name` - as I had expected from the opening post.
Anyway, I can add, edit and save the value.

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/12b8100d-58b5-431d-8bcd-6b88292d3713)

I have also tried the "bread" example. It worked well when I found out that the request requires the variant id, not the product id. I found this misleading because the URL actually has "supplied_product" in it - not sure if this could benefit from some clarification.
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/d60484ad-309c-43d8-bc21-b136b7b58b85)

In the end it worked well. :+1: 
Thanks again! I will merge this one! :rocket:
